back/src/main/java/com/lab/elephant/controller/PermissionController.java

```diff
@@ -64,15 +62,74 @@ public void add(@RequestBody ShareNoteDTO dto, @PathVariable long noteId) {
   
   @GetMapping(path = "{noteId}/permission")
   public String getPermission(@PathVariable long noteId) {
-    final Optional<Note> optionalNote = noteService.getNote(noteId);
-    final String string = (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
-    final User user = userService.getByEmail(string).get();
-    if (!optionalNote.isPresent())
-      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Note Not Found");
-    Note note = optionalNote.get();
+    final Note note = getNote(noteId);
+    final User user = getUser();
     final List<User> usersWithPermissions = noteService.getUsersWithPermissions(note);
     if (!usersWithPermissions.contains(user))
       throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "User has no Permission");
-    return permissionService.getPermissionBetween(user, note).get().toString();
+    return permissionService.getPermissionTypeBetween(user, note).get().toString();
+  }
+  
+  @GetMapping(path = "/allPermissions/{noteId}")
+  public List<PermissionDTO> getAllPermissions(@PathVariable long noteId) {
+    final Note note = getNote(noteId);
+    final User user = getUser();
+    checkOwnership(user, note);
+    final List<PermissionDTO> list = new ArrayList<>();
+    for (Permission p : note.getPermissions())
+      if (!p.getType().equals(PermissionType.Owner))
+        list.add(new PermissionDTO(p));
+    
+    return list;
+  }
+  
+  @PutMapping(path = "editPermissions/{noteId}")
+  public void editPermissions(@PathVariable long noteId, @RequestBody EditPermissionDTO dto) {
+    final Note note = getNote(noteId);
+    final User user = getUser();
+    checkOwnership(user, note);
+  
+    final List<PermissionDTO> list = dto.getList();
+  
+    for (PermissionDTO e : list) {
```

Comment:
El update debería hacerse para todos los permisos o para ninguno. En teoría se podría hacer solo para los que están bien, pero hace complicado saber cuales se cambiaron y cuales no para mostrarlo en el frontend (si no se devuelve nada). 

El problema con como esta hecho ahora es que se editan solo los correctos hasta el primer incorrecto, por lo que también hay algunos permisos correctos que no se editan. No tiene mucho sentido que se devuelva `BAD_REQUEST` pero que algunos se actualicen, porque esa respuesta da a entender que no se pudo procesar y que por lo tanto no pasó nada.

No hay que hacer tantos cambios para que funcione bien. Lo más fácil creo que es primero hacer toda la parte de verificación (sería sacar el `editRelationship()` del `for`). Si alguno falla, va a tirar la excepción correspondiente pero nada se va a actualizar. Lo que podrías hacer con los usuarios a los que se les van a cambiar los permisos es guardarlos en una lista en el mismo orden que están en la request (porque los necesitas para la segunda parte). Una vez que verificaste todos los permisos de la request, harías un nuevo `for` que los actualice todos.

Otra cosa que hay que fijarse es que ninguno de los usuarios a los que se les quiere cambiar su permiso sea el owner de la nota (sería fijarse que el `PermissionType` no sea `Owner`). Ahora se verifica que no se le de permiso de owner a otro pero no que se le saque al owner. Eso hay que verificarlo porque una nota sin owner traería problemas en otras partes.